### ISSUE TITLE
Cannot reference interfaces instances from an inflight context

### ISSUE THREAD

**Issue by ekeren** [winglang/wing#1961]
### Community Note

> Please vote by adding a 👍 reaction to the issue to help us prioritize.
> If you are interested to work on this issue, please leave a comment.

### I tried this:

```ts (wing)
bring cloud;

interface IKVStore {
  inflight get(k: str): str;
  inflight set(k: str, v: str);
}

resource BucketStore impl IKVStore {
  bucket: cloud.Bucket;  
  init() {
    this.bucket = new cloud.Bucket();
  }
  
  inflight get(k: str): str {
    return this.bucket.get(k);
  }
  
  inflight set(k: str, v: str){
    this.bucket.put(k, v);
  }

}

let kv: IKVStore = new BucketStore();

new cloud.Function(inflight () => {
  kv.set("key", "value");
  assert("value" == kv.get("key"));
}) as "test";
```

### I expected this:

compile and pass the test

### Instead, this happened:

```sh
wing test app.w
error: Cannot reference 'kv' of type 'IKVStore' from an inflight context
   --> app.w:27:3
   |
27 |   kv.set("key", "value");
   |   ^^ Cannot reference 'kv' of type 'IKVStore' from an inflight context


error: Cannot reference 'kv' of type 'IKVStore' from an inflight context
   --> app.w:28:21
   |
28 |   assert("value" == kv.get("key"));
   |                     ^^ Cannot reference 'kv' of type 'IKVStore' from an inflight context

```

### Component:

Compiler

### Wing Version:

0.8.23

**Comment by yoav-steinberg** [comment]
The problem here is that the compiler can't know that `kv` is a resource. I think the best way to handle this is to declare that the interface is an interface of a resource. Current the code almost supports this:
```wing
interface IKVStore extends core.IResource {
  inflight get(k: str): str;
  inflight set(k: str, v: str);
}
```
This would work except that `core.IResource` isn't exposed by the sdk. Here are some alternatives:
1. Move `core.IResource` to `cloud.IResource` and expose it, then make `IKVStore` extend it.
2. Have some syntax for explicitly defining a resource interface:
```wing
resource interface IKVStore { ... }
```

**Comment by eladb** [comment]
Yes, I believe that this is something @Chriscbr had already handled somewhere (I think around the function handler interfaces). The [`is_capturable`](https://github.com/winglang/wing/blob/60bee796d5368256b476ed25fd9dcc854f1a7a35/libs/wingc/src/type_check.rs#L788) function already returns `true` if this is a "resource interface".

I like the `resource interface` syntax, which aligns with my desire for us to change `resource` to `resource class`.

However, I am okay for now to just export `core.IResource` and require folks to extend it.

**Comment by MarkMcCulloh** [comment]
+1 for the `resource` modifier. I like that it adds a consistent restraint to both interfaces and classes. It's similar to the consideration of a `json` modifier keyword so it fits with a theme that's going on.

**Comment by ekeren** [comment]
If we make the `resource class` change I suggest we make ASAP so we can "feel" it before the beta

I think that @yoav-steinberg initial reaction was against the `resource` modifier in the team meeting

**Comment by staycoolcall911** [comment]
Discussed this with @yoav-steinberg now - we're both comfortable with the `resource` modifier. Let's go with this direction.

**Comment by ekeren** [comment]
We decided that this ticket https://github.com/winglang/wing/issues/2111 is P1
While the current Ticket is P2

But in practice, this issue actually means that you cannot really use preflight classes as interfaces (unless you only need to use them only in preflight)

I would argue that is is better to leave interfaces in wing out of the scope of the beta and not cause this confusion.

**Comment by Chriscbr** [comment]
@ekeren Is this still an issue? I updated the example to the latest Wing and I think it's working

```js
bring cloud;

interface IKVStore {
  inflight get(k: str): str;
  inflight set(k: str, v: str): void;
}

class BucketStore impl IKVStore {
  bucket: cloud.Bucket;  
  new() {
    this.bucket = new cloud.Bucket();
  }
  
  pub inflight get(k: str): str {
    return this.bucket.get(k);
  }
  
  pub inflight set(k: str, v: str){
    this.bucket.put(k, v);
  }

}

let kv: IKVStore = new BucketStore();

test "inflight access to interface" {
  kv.set("key", "value");
  assert("value" == kv.get("key"));
}
```
